### DISCUSSION TITLE
HigherOrderGP Posterior eval mode

### DISCUSSION THREAD

**Discussion by ChristianMichelsen** [meta-pytorch/botorch#2746] [General]
In HigherOrderGP, once you call `.posterior` it automatically sets it in `eval` mode, but does not set it back to `train` mode once it's finished. Is there are reason why it does not do that?

I've had a couple of times where I initialize a HOGP and just quickly check predictions at first before optimizing the MLL to be able to see the progress but without any progress due to the initial call to `.posterior`. This had left me spending more time on debugging more than I am proud of 😅

**Comment by saitcakmak** [comment]
Hi @ChristianMichelsen. We generally do not put models in `.train` outside of model training. Calling `train()` on a GPyTorch `Module` will trigger a `self._clear_cache()` call, which will destroy important caches that make repeated model inference efficient.

**Comment by ChristianMichelsen** [comment]
Would it then make sense to put  the gp into `.train()` mode in e.g. `fit_gpytorch_mll_scipy`?

Also, regarding the caches, where can I read more about them, that sounds interesting!

**Reply by saitcakmak** [reply]
> Would it then make sense to put the gp into .train() mode in e.g. fit_gpytorch_mll_scipy?

We do this in `fit_gpytorch_mll`, the main entry point for model fitting, which accepts `fit_gpytorch_mll_scipy/torch` as the `optimizer` input. I don't think it'd hurt to do this in `fit_gpytorch_mll_scipy` as well, but I already see a note about this in the docstring:
> The model and likelihood in mll must already be in train mode.

> Also, regarding the caches, where can I read more about them, that sounds interesting!

I don't know if there is a good place to read about them. I'm mostly familiar with them through working with the code but you can check GPyTorch docs https://docs.gpytorch.ai/.
